storage/src/desktop/storage_reference_desktop.cc

```diff
@@ -279,18 +285,106 @@ Future<size_t> StorageReferenceInternal::GetBytes(void* buffer,
-      new GetBytesResponse(buffer, buffer_size, handle, future_api);
+// Sends a rest request, and creates a separate thread to retry failures.
+template <typename FutureType>
+void StorageReferenceInternal::SendRequestWithRetry(
+    StorageReferenceFn internal_function_reference,
+    SendRequestFunct send_request_funct,
+    SafeFutureHandle<FutureType> final_handle, double max_retry_time_seconds) {
+  BlockingResponse* first_response = send_request_funct();
+  std::thread async_retry_thread(
+      &StorageReferenceInternal::AsyncSendRequestWithRetry<FutureType>, this,
+      internal_function_reference, send_request_funct, final_handle,
+      first_response, max_retry_time_seconds);
+  async_retry_thread.detach();
+}
+
+const int kInitialSleepTimeMillis = 1000;
+const int kMaxSleepTimeMillis = 30000;
+
+// In a separate thread, repeatedly send Rest requests until one succeeds or a
+// maximum amount of time has passed.
+template <typename FutureType>
+void StorageReferenceInternal::AsyncSendRequestWithRetry(
+    StorageReferenceFn internal_function_reference,
+    SendRequestFunct send_request_funct,
+    SafeFutureHandle<FutureType> final_handle, BlockingResponse* response,
+    double max_retry_time_seconds) {
+  auto* future_api = future();
+  firebase::FutureBase internal_future;
+  auto end_time = std::chrono::steady_clock::now() +
+                  std::chrono::duration<double>(max_retry_time_seconds);
+  auto current_sleep_time = std::chrono::milliseconds(kInitialSleepTimeMillis);
+  auto max_sleep_time = std::chrono::milliseconds(kMaxSleepTimeMillis);
+  while (true) {
+    internal_future = future_api->LastResult(internal_function_reference);
+    // Wait for completion, then check status and error.
+    while (internal_future.status() == firebase::kFutureStatusPending) {
+      std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    }
+    // For any request that succeeds or fails in a non-retryable way, don't
+    // bother retrying.
+    if (internal_future.status() != firebase::kFutureStatusComplete ||
+        !IsRetryableFailure(response->status())) {
+      break;
+    }
+    // Interrupt the loop if the retry deadline has been reached
+    auto current_time = std::chrono::steady_clock::now();
+    if (current_time + current_sleep_time > end_time) {
+      break;
+    }
+    // Sleep for an exponentially increasing duration, then retry the request.
+    std::this_thread::sleep_for(current_sleep_time);
+    current_sleep_time *= 2;
+    if (current_sleep_time > max_sleep_time) {
+      current_sleep_time = max_sleep_time;
+    }
+    response = send_request_funct();
+  }
+  // Copy from the internal future to the final future.
+  Future<FutureType> typed_future =
+      static_cast<const Future<FutureType>&>(internal_future);
+  if (typed_future.result() != nullptr) {
+    if constexpr (std::is_void<FutureType>::value) {
+      future_api->Complete(final_handle, internal_future.error());
```

Comment:
Delete() returns a Future<void>. The code for Delete seemed to end up passing the result() != nullptr check but I initially encountered an error about not being able to get a reference to void, so I added an is_void check.

Another alternative would probably be to explicitly define a variation of this function for type void.

Reply:
I think this is fine, as long as it works on all 3 desktop platforms.